component/stream/stream_cache.go

```diff
@@ -200,3 +210,104 @@ func (c *cache) filePurge(key, value interface{}) {
 	c.blocks.Remove(key)
 	c.evictedBlock = blockKey{}
 }
+
+// Using key construct a file name for persisted block
+func (c *cache) getLocalFilePath(key blockKey) string {
+	return filepath.Join(c.diskPath, key.fileKey+"__"+fmt.Sprintf("%d", key.offset)+"__")
+}
+
+// Persist this block on disk
+func (c *cache) persistBlockOnDisk(block *cacheBlock, key blockKey) {
+	localPath := c.getLocalFilePath(key)
+
+	log.Debug("streamcache::persistBlockOnDisk : Saving file %s offset %d to disk", key.fileKey, key.offset)
+
+	err := os.MkdirAll(filepath.Dir(localPath), os.FileMode(0775))
+	if err != nil {
+		log.Err("streamcache::persistBlockOnDisk : unable to create local directory %s [%s]", localPath, err.Error())
+		return
+	}
+
+	f, err := os.OpenFile(localPath, os.O_RDWR|os.O_CREATE|os.O_TRUNC, 0775)
+	if err != nil {
+		log.Err("streamcache::persistBlockOnDisk : Failed to create file for persisting block on disk %s (%s)", localPath, err.Error())
+		return
+	}
+	f.Write(block.data)
+	f.Close()
+
+	c.diskBlocks.Set(key, true)
+}
+
+// Load data for this block from disk
+func (c *cache) getBlockFromDisk(block *cacheBlock, key blockKey) bool {
+	localPath := c.getLocalFilePath(key)
+	info, err := os.Stat(localPath)
+
+	if err != nil {
+		return false
+	}
+
+	log.Debug("streamcache::getBlockFromDisk : Reading block for %s offset %d from disk", key.fileKey, key.offset)
+	if time.Since(info.ModTime()).Seconds() > c.diskTimeoutSec {
+		// File exists on local disk but disk cache timeout has elapsed
+		os.Remove(localPath)
+		c.diskBlocks.Remove(key)
+		return false
+	}
+
+	f, err := os.OpenFile(localPath, os.O_RDONLY, 0775)
+	if err != nil {
+		return false
+	}
+
+	f.Read(block.data)
+	f.Close()
+	os.Remove(localPath)
+
+	c.diskBlocks.Remove(key)
+
+	return true
+}
+
+// Remove this block from in-memory cache
+func (c *cache) evictBlock(key blockKey, block *cacheBlock) {
```

Comment:
There is one very important unit test we should have, here consider the following:
1. our number of cached files is full
2. the cache evicts one of the files
3. this causes a purge on all blocks in memory for that file
4. this in turn persists these blocks on disk (while the fileKey and its blocks are now gone from memory)
5. another read to that evicted file comes for one of the blocks on disk
6. we need to make sure that the key is handled correctly and we end up fetching the block from disk rather than getting another copy of it (leading to duplicate data on disk with a stale block from the last file purge)

I know we're using key.offset but just for future proofing this test is important to have to make sure no memory leaks are happening since currently its event driven

Reply:
There is a test in UT where we bring the block in memory and then create a eviction case where it goes back to disk. Later when the block is read its read from disk. This is similar to cache full scenario as in that UT we take cache size to be 1 block per file and simulate the UT around it.